### PR TITLE
Re-use printer for multiple elements

### LINKED ISSUE #2091: Tkn bundle list split yamls
### Feature request

It would be a useful feature for us when we use tkn bundle list on an OCI image, that contains multiple tasks/pipelines the output would be a valid merged yaml file. Right now it does not separate the yamls with ---. Therefor if we use tkn bundle list, then pipe it to kubectl apply it only applies the last task/pipeline, not all of them.

### Use case

If it would be implemented we could use bundle to bundle a Pipeline with all of its Tasks into one OCI Image, then we could version control our Pipelines and could easily deploy them from CI with one tkn bundle list piped into kubectl apply.

### PR DESCRIPTION
**Pull request by lcarva**
# Changes

Some printers contain an internal state. The YAMLPrinter, for example, tracks how many objects have been printed. When it prints an object, and it is not the first one to be printed, it prints the YAML document separator, `---`. If a new printer is used when printing each object, then the YAMLPrinter always considers that it is printing the first object. This causes the YAML objects to be muddled together.

This commit makes it so a printer is re-used when printing multiple objects. For conformity, this is also done when a single object is expected to be printed, e.g. get or describe.

Fixes #2091

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Run the code checkers with `make check`
- [x] Regenerate the manpages, docs and go formatting with `make generated`
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```release-note
Fix printing of multiple YAML documents
```

### CODE REVIEW

File: pkg/cmd/bundle/list.go [hunk 1]
```diff
@@ -113,8 +112,13 @@ Caching:
 				Err: cmd.OutOrStderr(),
 			}
 
+			p, err := f.ToPrinter()
```
Comment (piyush-garg) [inline pkg/cmd/bundle/list.go:115]: can we keep this also inside `PrintObjects` function, it is getting duplicated at multiple places

File: pkg/cmd/triggertemplate/list.go [hunk 2]
```diff
@@ -87,7 +86,11 @@ or
 			}
 
 			if output != "" {
-				return printer.PrintObject(stream.Out, tts, f)
+				p, err := f.ToPrinter()
```
Comment (piyush-garg) [inline pkg/cmd/triggertemplate/list.go:89]: can we keep this also as function like before, it is getting duplicated at multiple places

File: pkg/cmd/bundle/list.go (hunk 1 above)
Comment (lcarva) [reply pkg/cmd/bundle/list.go:115]: Notice how `PrintObjects` was called for each item in the list, individually. This means a new instance of the printer is created for each call which causes the issue in the first place. I tried to setup some sort of caching where the printer would only be created once per output format. The unit tests revealed that to be a problematic pattern as it forces a global state between different commands.

Any suggestions?

Comment (piyush-garg) [reply pkg/cmd/bundle/list.go:115]: ok this makes sense, for now will go with this

File: pkg/cmd/triggertemplate/list.go (hunk 2 above)
Comment (piyush-garg) [reply pkg/cmd/triggertemplate/list.go:89]: this comment is invalid based on above comment reply

**Review by vinamra28 — APPROVED** [review]

**Pull request merged**